packages/ai/src/ui/chat-store.ts

```diff
@@ -0,0 +1,528 @@
+          );
+          existingToolInvocation.args = partialArgs;
+          break;
+        }
+        case 'call': {
+          existingToolInvocation.args = toolInvocation.args;
+          break;
+        }
+        case 'result': {
+          // @ts-expect-error - Due to the switch statement narrowing down the type of `existingToolInvocation`, TS is not aware that toolInvocation is also a result Tool Invocation even though both are the same type:
+          existingToolInvocation.result = toolInvocation.result;
+          break;
+        }
+        default: {
+          throw new Error('Invalid tool invocation state');
+        }
+      }
+    } else {
+      if (toolInvocation.state === 'result') {
+        throw new Error('tool_result must be preceded by a tool_call');
+      }
+
+      if (
+        toolInvocation.state !== 'partial-call' &&
+        toolInvocation.state !== 'call'
+      ) {
+        throw new Error('Invalid tool invocation state');
+      }
+
+      const partialToolPart: ToolInvocationUIPart = {
+        type: 'tool-invocation',
+        toolInvocation: {
+          args: toolInvocation.args ?? '',
+          step: this.calculateStep(chatId),
+          toolName: toolInvocation.toolName,
+          state: toolInvocation.state,
+          toolCallId,
+        } as ToolInvocation,
+      };
+
+      partialState.toolParts[toolCallId] = partialToolPart;
+      partialState.toolArgs[toolCallId] = toolInvocation.args ?? '';
+      message.parts.push(partialToolPart);
+    }
+  }
+
+  private addOrUpdateReasoning({
+    reasoning,
+    activeResponse,
+  }: {
+    reasoning: ReasoningUIPart;
+    activeResponse: ChatState['activeResponse'];
+  }) {
+    if (!activeResponse) return;
+
+    const partialReasoning = activeResponse.partialState.reasoningPart;
+
+    const { text: reasoningTextDelta, providerMetadata } = reasoning;
+
+    const isFinishedPart =
+      partialReasoning &&
+      reasoningTextDelta === '' &&
+      providerMetadata === undefined;
+
+    // Clear temporary state on `reasoning_part_finish`:
+    if (isFinishedPart) {
+      activeResponse.partialState.reasoningPart = undefined;
+      return;
+    }
+
+    if (partialReasoning) {
+      partialReasoning.text += reasoningTextDelta;
+    } else {
+      activeResponse.partialState.reasoningPart = {
+        type: 'reasoning',
+        text: reasoningTextDelta,
+      };
+      activeResponse.message.parts.push(
+        activeResponse.partialState.reasoningPart,
+      );
+    }
+
+    // Only update provider metadata if defined in the reasoning part:
+    if (providerMetadata) {
+      activeResponse.partialState.reasoningPart!.providerMetadata =
+        providerMetadata;
+    }
+  }
+
+  clear(id?: string) {
+    if (id) {
+      this.setMessages({ id, messages: [] });
+      this.resetActiveResponse(id);
+    } else {
+      const ids = Array.from(this.chats.keys());
+      for (const id of ids) {
+        this.clear(id);
+      }
+    }
+  }
```

Comment:
why not just evict from map? if not, error also needs to be cleared.

needs to abort ongoing streaming?